2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,231 @@
+
+If the module were named `PersonFormatter`, it might make more sense to name the function something else, perhaps simply `name`. Naming the module is outside the purview of this article, but will be addressed in a followup article.
+
+Diagram 2: Function names abstraction ladder
+
+<img src="images/function-names-ladder.svg" width="100%" alt="Function Names Abstraction Ladder" />
+
+Should you need to change how the name is formatted in response to upstream data changes or new application requirements, the impact of the implementation will be constrained to the internals of the function. The name will continue to be appropriate and no dependents will be forced to change.
+
+In terms of the Open Closed Principle, the code is more open to the change. For example, a new requirement to add a salutation would only necessitate changes to implementation internals, no API-breaking name change would be necessary. By increasing the level of abstraction, you have created an approprate and resilient function name.
+
+**GUIDELINE: Name functions one level of abstraction higher than their implementation**
+
+### Climb higher: ask a friend for stability
+
+When you seriously consider function names, it's not uncommon to come up with several candidates that seem equally viable. In those cases, break the tie by choosing words that your customers and product team use. In other words, when you're not sure, ask others for help. Just as you might ask a friend to stabilize the ladder while you climb that one rung higher to reach the work you need to do, asking for help in naming has a stabilizing effect on code.
+
+Application lexicons arise from conversations. It reduces mental friction when the vocabulary of your team matches the words you see on the screen in your text editor. Without having to translate person-speak to code-speak, you remove just one more bit of mental processing you'd do best to avoid. Agreeing on names reduces the overall organizational cost of writing software. In this spirit, it can be helpful, especially for newcomers on your team, to maintain a glossary of terms for your application.
+
+**GUIDELINE: Choose names from your application lexicon**
+
+## Climb up and down the ladder
+
+Does software design really demand this level of attention to detail, this careful weighing of names for something as unremarkable and ordinary as a single function? If you want to do your best work, yes, it does. Names matter. Attention to details matters. Writing software is about more than just getting something to work. At an abstract level, it's an expression of thought and communication, like writing, like speaking, like relationship-building. Your thoughtfulness around names will pay off.
```

Comment:
Maybe needs "then":  "If you want to do your best work, then yes, it does."

Reply:
Resolved in 345a1339f8f2b4e506e3897474916427967dd070